Implement requestHeaders and requestData processing in fetch library

quyse opened the pull request:
Implementation for actual sending of `requestHeaders` and `requestData` fields from `emscripten_fetch_attr_t` structure in the `fetch` library.

The patch so far is very simple, I just needed this feature for my current project.

juj commented:
Thanks, this looks good. Sorry for the delay in getting to review. Would you be able to rebase up to tip? Also, there is one change that is needed - the code needs to prepare for the case that `malloc()` fails to allocate and returns 0. In such case, it would be good to `free()` all temp allocations and return failure from the call. (I know it's been a while so let me know, I can also do that change if you have already moved on to tackling other problems?)

quyse commented:
Hi @juj, thanks for looking at this! I rebased the patch onto latest `incoming`, and added a few checks for out of memory conditions as you requested. I also added a few missing `free`s, fixing those TODO comments. I had no time yet to properly test these new changes, sorry. I hope l'll look at this again over weekend.

juj commented on system/lib/fetch/emscripten_fetch.cpp:
```diff
@@ -67,16 +67,49 @@ emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const
 	}
 
 	emscripten_fetch_t *fetch = (emscripten_fetch_t *)malloc(sizeof(emscripten_fetch_t));
+	if (!fetch) return 0;
```
Good catch on this existing alloc failing.

juj commented on system/lib/fetch/emscripten_fetch.cpp:
```diff
@@ -67,16 +67,49 @@ emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const
+	fetch->__attributes.overriddenMimeType = fetch->__attributes.overriddenMimeType ? strdup(fetch->__attributes.overriddenMimeType) : 0;
+	if (fetch->__attributes.requestData && fetch->__attributes.requestDataSize)
+	{
+		char* data = (char*)malloc(fetch->__attributes.requestDataSize);
```
Ouch, this line makes me regret: when POSTing to upload data, the data size can potentially be very large, so it is possible that this `malloc` is on the order of dozens to even up to ~hundred of megabytes in size, and could imaginably fail more easily than the other allocs, since while the fetch is happening, this means we'll be temporarily creating two copies of the uploadable data into the Wasm HEAP.

That makes me think that we should define the semantics of the `requestData` pointer specially that the caller should keep that pointer alive throughout the duration of the call, or otherwise large uploads can run into trouble.

I see that I have documented the `requestData` field as follows in `fetch.h`:

```c
	// If non-zero, specifies a pointer to the data that is to be passed as the body (payload) of the request
	// that is being performed. Leave as zero if no request body needs to be sent.
	// The memory pointed to by this field is provided by the user, and needs to be valid only until the call to
	// emscripten_fetch() returns.
	const char *requestData;
```

however I never had time to implement this field originally, so I think we still have a chance to fix this and make `requestData` be special since it's a potentially fat block of data:

```c
	// If non-zero, specifies a pointer to the data that is to be passed as the body (payload) of the request
	// that is being performed. Leave as zero if no request body needs to be sent.
	// The memory pointed to by this field is provided by the user, and needs to be valid throughout the
        // duration of the fetch operation. If passing a non-zero pointer into this field, make sure to implement
        // the onsuccess and onerror handlers to be notified when the fetch finishes to know when this memory
        // block can be freed. Do not pass a pointer to memory on the stack or other temporary area here.
	const char *requestData;
```

What do you think?

juj commented on system/include/emscripten/fetch.h:
```diff
@@ -172,6 +172,9 @@ EMSCRIPTEN_RESULT emscripten_fetch_wait(emscripten_fetch_t *fetch, double timeou
 EMSCRIPTEN_RESULT emscripten_fetch_close(emscripten_fetch_t *fetch);
 
+// For internal use only. Frees all memory related to fetch operation.
+void emscripten_fetch_free(emscripten_fetch_t *fetch);
```
Let's not make this public, the idea was that `emscripten_fetch_close()` would be the function that frees, so let's delete this from here.

juj commented on system/lib/fetch/emscripten_fetch.cpp:
```diff
@@ -144,8 +177,26 @@ EMSCRIPTEN_RESULT emscripten_fetch_close(emscripten_fetch_t *fetch)
+	return EMSCRIPTEN_RESULT_SUCCESS;
+}
+
+void emscripten_fetch_free(emscripten_fetch_t *fetch)
```
Good idea to share the implementation out here, but to make this private, let's rename to `static void _emscripten_fetch_free(emscripten_fetch_t *fetch)` so that it is local only to this cpp file?

juj commented on system/lib/fetch/emscripten_fetch.cpp:
```diff
@@ -67,16 +67,49 @@ emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const
+		size_t headersCount;
+		for(headersCount = 0; fetch->__attributes.requestHeaders[headersCount]; ++headersCount);
+		const char** headers = (const char**)malloc((headersCount + 1) * sizeof(const char*));
+		if(!headers)
```
The general guideline is to have a space after `if`, like also above, i.e. `if (`

froydnj commented on system/lib/fetch/emscripten_fetch.cpp:
```diff
@@ -67,28 +67,46 @@ emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const
 			headers[i] = strdup(fetch->__attributes.requestHeaders[i]);
+			if(!headers[i])
+			{
+				emscripten_fetch_free(fetch);
```
Note that if one of the `strdup` calls (or even the allocation of `headers`, above) fails, `emscripten_fetch_free` is going to attempt to free `fetch->__attributes.requestHeaders`, which is what was passed in via `fetch_attr`, not the `requestHeaders` being allocating here.  This behavior is probably not the right thing.

Also, if the `strdup` calls fail, `headers` itself and any other strings that have been duplicated also need to be freed.

juj replied:
If `strdup` fails, it should gracefully return 0 if I understand correctly. But when that happens, we indeed must make sure we have `strdup`ped all the fields, or zero them out manually so that we don't go and free user-submitted pointers. That's a very good catch.

Related to this, I was thinking that if one of the `strdup`s fail, it will mean that the fetch can fire off in "partial" state, with one or more of the strdupped fields not being present in the initiated fetch. That's something I'm thinking is something that can be left for a followup to address, since it will be a bit more complex, and I don't want to "string in" a collaborator to start to do extra work to fix everything in one go.

quyse commented:
Thanks for reviewing @juj and @froydnj! I hope I applied all the changes requested, and also rebased again onto latest `incoming`. Absolutely agree that POST data must not be copied. I added checks to all calls to `strdup` and tried to reduce copy-pasting by introducing local macro (not sure if you'd like that, but don't see better way). And finally I was able to restore working emscripten environment on my PC and do some testing of these changes with a real world application, seems OK so far.

abcthomas commented:
Is this actively being looked at @juj  ? It would be extremely helpful and useful as the Fetch API can't really be used without it.

kripken commented:
The main blocker here is probably that @juj doesn't have much time to review code these days. Aside from that, reading this and the parallel PR #6207, it sounds like we need to add the test from there to here. Once that's done, I can try to review the code myself in order to speed up landing this.

abcthomas commented:
Yeah that's fair I understand that. Ok thanks that would be really helpful, @quyse are you able to transfer the test code from PR #6207 ? If the original poster doesn't reply/is no longer around, then I might combine the two on my own branch and do a new PR

quyse commented:
@abcthomas Sorry for my absence, at the moment I don't have time to work on this. If you can combine this with tests in your PR that's great!

kripken commented:
This work has moved to #6514, which was merged, closing.
